utils/yoke/derive/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+                    mem::forget(this);
+                    ptr::read(ptr)
+                }
+                fn transform_mut<F>(&'a mut self, f: F)
+                where
+                    F: 'static + for<'b> FnOnce(&'b mut Self::Output) {
+                    unsafe { f(std::mem::transmute::<&'a mut Self, &'a mut Self::Output>(self)) }
+                }
+            }
+        }
+    }
+}
+
+/// Custom derive for `yoke::ZeroCopyFrom`,
+///
+/// This implements `ZeroCopyFrom<Ty> for Ty` for types
+/// without a lifetime parameter, and `ZeroCopyFrom<Ty<'s>> for Ty<'static>`
+/// for types with a lifetime parameter.
+///
+/// Apply the `#[yoke(CloningZCF)]` attribute if you wish for this custom derive
+/// to use `.clone()` for its implementation.
+#[proc_macro_derive(ZeroCopyFrom, attributes(yoke))]
+pub fn zcf_derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    TokenStream::from(zcf_derive_impl(&input))
+}
+
+fn zcf_derive_impl(input: &DeriveInput) -> TokenStream2 {
+    let typarams = input.generics.type_params().count();
+    if typarams != 0 {
+        return syn::Error::new(
+            input.generics.span(),
+            "derive(ZeroCopyFrom) does not support type parameters",
+        )
+        .to_compile_error();
+    }
+    let has_clone = input.attrs.iter().any(|a| {
+        if let Ok(i) = a.parse_args::<Ident>() {
+            if i == "CloningZCF" {
+                return true;
+            }
+        }
+        false
+    });
+    let lts = input.generics.lifetimes().count();
+    let name = &input.ident;
+    if lts == 0 {
+        let clone = if has_clone {
+            quote!(this.clone())
+        } else {
+            quote!(*this)
+        };
+        quote! {
+            impl ZeroCopyFrom<#name> for #name {
+                fn zero_copy_from(this: &Self) -> Self {
+                    #clone
+                }
+            }
+        }
+    } else {
+        if lts != 1 {
+            return syn::Error::new(
+                input.generics.span(),
+                "derive(ZeroCopyFrom) cannot have multiple lifetime parameters",
+            )
+            .to_compile_error();
+        }
+        if has_clone {
+            return quote! {
+                impl<'s> ZeroCopyFrom<#name<'static>> for #name<'static> {
+                    fn zero_copy_from(this: &Self) -> Self {
+                        this.clone()
+                    }
+                }
+            };
+        }
+
+        let structure = Structure::new(&input);
+        let body = structure.each_variant(|vi| {
+            vi.construct(|f, i| {
+                let binding = format!("__binding_{}", i);
+                let field = Ident::new(&binding, Span::call_site());
+                let fty = replace_lifetime(&f.ty, static_lt());
+                quote! {
+                    <#fty as yoke::ZeroCopyFrom<_>>::zero_copy_from(#field)
+                }
+            })
```

Comment:
Thought: I think this is the important part, the one that requires `ZeroCopyFrom` to be implemented on all the fields.

Reply:
Correct!